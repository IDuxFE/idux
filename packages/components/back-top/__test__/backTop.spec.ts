import { mount } from '@vue/test-utils'
import IxBackTop from '../src/BackTop.vue'
import { renderWork, waitRAF, isShow, wait, scrollTarget } from '@tests'
import { nextTick } from 'vue'

const mockFn = jest.fn()
const warn = jest.spyOn(console, 'warn').mockImplementation()

const backTopMount = (template: string, options = {}) =>
  mount(
    {
      components: { IxBackTop },
      template,
      ...options,
    },
    { attachTo: document.body },
  )

describe('BackTop', () => {
  renderWork(IxBackTop)

  test('scroll work', async () => {
    const wrapper = backTopMount(
      `
      <div class="ix-back-top-test" style="height: 300px; overflow: auto;">
        <div style="height: 1000px;">
          <ix-back-top :duration="100" :visibilityHeight="200" target=".ix-back-top-test" />
        </div>
      </div>
    `,
      {
        beforeUnmount() {
          mockFn()
        },
      },
    )

    expect(isShow(wrapper.find('.ix-back-top'))).toBe(false)

    await scrollTarget(600, wrapper.element)

    expect(isShow(wrapper.find('.ix-back-top'))).toBe(true)

    await wrapper.find('.ix-back-top').trigger('click')
    await wait(200)

    expect(wrapper.element.scrollTop).toBe(0)

    wrapper.unmount()
    expect(mockFn).toBeCalled()
  })

  test('props work: target is a HTMLElement', async () => {
    const wrapper = backTopMount(
      `
      <div style="height: 1000px;">
        <ix-back-top :target="target" />
      </div>
    `,
      {
        data() {
          return {
            target: document.documentElement,
          }
        },
      },
    )

    expect(isShow(wrapper.find('.ix-back-top'))).toBe(false)

    await scrollTarget(600, document.documentElement)

    expect(isShow(wrapper.find('.ix-back-top'))).toBe(true)
  })

  test('props work: target does not exist', async () => {
    mount(IxBackTop, { props: { target: '#ix-back-top-test' } })

    await nextTick()

    expect(warn).toBeCalled()
  })

  test('props work: target is the default value', async () => {
    window.scrollTo = mockFn

    const wrapper = backTopMount(`
      <div style="height: 1000px; overflow: auto;">
        <ix-back-top />
      </div>
    `)

    await scrollTarget(600, window)
    await wrapper.find('.ix-back-top').trigger('click')
    await waitRAF()

    expect(mockFn).toBeCalled()
  })
})
