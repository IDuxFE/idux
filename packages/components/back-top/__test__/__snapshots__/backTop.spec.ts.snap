// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackTop.vue render work 1`] = `
"<transition-stub>
  <div class=\\"ix-back-top\\" style=\\"display: none;\\"><i class=\\"ix-icon ix-icon-vertical-align-top\\" role=\\"img\\" arialabel=\\"vertical-align-top\\"></i></div>
</transition-stub>"
`;
